Code/Framework/AzFramework/Platform/Linux/AzFramework/Input/Devices/Keyboard/InputDeviceKeyboard_xcb.cpp

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/std/reference_wrapper.h>
+#include <AzCore/std/sort.h>
+#include <AzCore/std/typetraits/integral_constant.h>
+#include <AzFramework/API/ApplicationAPI_Linux.h>
+#include <AzFramework/Input/Buses/Notifications/RawInputNotificationsBus_xcb.h>
+#include <AzFramework/Input/Devices/Keyboard/InputDeviceKeyboard.h>
+
+#define explicit ExplicitIsACXXKeyword
+#include <xcb/xkb.h>
+#undef explicit
+#include <xkbcommon/xkbcommon-keysyms.h>
+#include <xkbcommon/xkbcommon.h>
+#include <xkbcommon/xkbcommon-x11.h>
+
+namespace AzFramework
+{
+    class InputDeviceKeyboardXcb
+        : public InputDeviceKeyboard::Implementation
+        , public RawInputNotificationBusXcb::Handler
+    {
+    public:
+        AZ_CLASS_ALLOCATOR(InputDeviceKeyboardXcb, AZ::SystemAllocator, 0);
+
+        using InputDeviceKeyboard::Implementation::Implementation;
+        InputDeviceKeyboardXcb(InputDeviceKeyboard& inputDevice)
+            : InputDeviceKeyboard::Implementation(inputDevice)
+        {
+            RawInputNotificationBusXcb::Handler::BusConnect();
+
+            auto* connection = AzFramework::LinuxXcbConnectionManagerInterface::Get()->GetXcbConnection();
+
+            AZStd::unique_ptr<xcb_xkb_use_extension_reply_t, DeleterForFreeFn<::std::free>> xkbUseExtensionReply{
+                xcb_xkb_use_extension_reply(connection, xcb_xkb_use_extension(connection, 1, 0), nullptr)
+            };
+            m_coreDeviceId = xkb_x11_get_core_keyboard_device_id(connection);
+
+            m_xkbContext.reset(xkb_context_new(XKB_CONTEXT_NO_FLAGS));
+            m_xkbKeymap.reset(xkb_x11_keymap_new_from_device(m_xkbContext.get(), connection, m_coreDeviceId, XKB_KEYMAP_COMPILE_NO_FLAGS));
+            m_xkbState.reset(xkb_x11_state_new_from_device(m_xkbKeymap.get(), connection, m_coreDeviceId));
+        }
+
+        bool IsConnected() const override
+        {
+            auto* connection = AzFramework::LinuxXcbConnectionManagerInterface::Get()->GetXcbConnection();
+            return connection && !xcb_connection_has_error(connection);
+        }
+
+        bool HasTextEntryStarted() const override
+        {
+            return false;
+        }
+
+        void TextEntryStart(const InputDeviceKeyboard::VirtualKeyboardOptions& options) override
+        {
+        }
+
+        void TextEntryStop() override
+        {
+        }
+
+        void TickInputDevice() override
+        {
+            ProcessRawEventQueues();
+        }
+
+        void OnRawInputEvent(xcb_generic_event_t* event) override
+        {
+            switch (event->response_type & ~0x80)
```

Comment:
If there is no constant, maybe a comment explaining the magic '0x80'?

Reply:
~0x80 is the new 0x7f :-)

https://github.com/o3de/o3de/pull/3975

Reply:
`switch` statements also support initialization in C++17 as well.
Therefore a constant be can to explain the event response mask

```suggestion
            switch (auto eventResponseMask = event->response_type & ~0x80; eventResponseMask )
```

Reply:
I found this documented in the [X Window System Protocol on events](https://www.x.org/releases/current/doc/xproto/x11protocol.html#event_format):
> Every event contains an 8-bit type code. The most significant bit in this code is set if the event was generated from a SendEvent request.

So you have to turn that bit off to get the event code.